logic/scorer/scorer.go

```diff
@@ -60,260 +34,247 @@ func points(lead model.Card, hand []model.Card, isCrib bool) int {
+	if (ptVals[0]|ptVals[1]|ptVals[2]|ptVals[3]|ptVals[4])&1 == 0 {
+		// all even numbered cards => no fifteens possible
+		return 0, 0
 	}
 
-	numPairs := 0
-	for i := 1; i < len(values); i++ {
-		if values[i] == values[i-1] {
-			numPairs++
-		}
-	}
-	if numPairs == 2 {
-		return twopair, 4 /* two points per pair*/
+	sum := ptVals[0] + ptVals[1] + ptVals[2] + ptVals[3] + ptVals[4]
+	if sum == 15 {
+		// only one fifteen possible
+		return fifteen1, 2
+	} else if sum < 15 || sum > 46 {
+		return 0, 0
 	}
 
-	if values[3]-min == 2 {
-		// either two pair or a double run of three
-		// but we already checked two pair
-		return doubleRunOfThree, 8 /* double run of 3 */
-	} else if max-values[1] == 2 {
-		// either two pair or a double run of three
-		// but we already checked two pair
-		return doubleRunOfThree, 8 /* double run of 3 */
+	var numFifteens int
+
+	for i := 0; i < len(ptVals); i++ {
+		many := howManyAddUpTo(15-ptVals[i], ptVals[i+1:])
+		numFifteens += many
 	}
 
-	var st scoreType
-	pts := 0
+	if numFifteens > 0 {
+		return fifteen0 << numFifteens, numFifteens * 2
+	}
+	return 0, 0
+}
 
-	if ((min+1 == values[1]) && (values[1]+1 == values[2])) ||
-		((values[2]+1 == values[3]) && (values[3]+1 == max)) {
-		st = st | run3 /* run of 3 (and maybe a pair) */
-		pts += 3
+func howManyAddUpTo(goal int, ptVals []int) int {
+	if len(ptVals) == 0 {
+		return 0
 	}
 
-	if numPairs == 1 {
-		// we already checked for two above
-		st = st | onepair /* one pair */
-		pts += 2
+	var many int
+	for i, o := range ptVals {
+		if o == goal {
+			many++
+		} else if o < goal {
+			// o is less than the goal. See what we can find with it
+			subWith := howManyAddUpTo(goal-o, ptVals[i+1:])
+			many += subWith
+		}
 	}
 
-	return st, pts
+	return many
 }
 
-func scoreFlushesAndNobs(lead model.Card, hand []model.Card, isCrib bool) (scoreType, int) {
-	var st scoreType
-	pts := 0
-	handSuit := hand[0].Suit
-	isHandFlush := true
-	for _, c := range hand {
-		if c.Suit != handSuit {
-			isHandFlush = false
+func scoreRuns(values [numCardsToScore]int, valuesToCounts []uint8) (scoreType, int) {
+	for _, v := range values {
+		if valuesToCounts[v-1] > 0 {
+			// this is already part of a run; skip calculation
+			continue
 		}
-		if c.Value == 11 /* Jack */ && c.Suit == lead.Suit {
-			st = st | nobs
-			pts++
+		runLen := uint8(1)
+		mult := valuesToCounts[v]
+		// we're at the potential beginning of a run
+		for next := v + 1; valuesToCounts[next] > 0; next++ {
+			mult *= valuesToCounts[next]
+			runLen++
+		}
+		if runLen >= 3 {
+			return calculateTypeAndPoints(runLen, mult)
 		}
 	}
+	return 0, 0
+}
 
-	if isHandFlush {
-		if lead.Suit == handSuit {
-			st = st | flush5
-			pts += 5
-		} else if !isCrib {
-			st = st | flush4
-			pts += 4
-		}
+func calculateTypeAndPoints(longest, mult uint8) (scoreType, int) {
+	if longest == 5 {
+		return run5, 5
 	}
+	// typeMap maps the run length and multiplier to a scoring type
+	typeMap := [5][5]scoreType{
```

Comment:
:mag:  can `typeMap` (have a new name and) be a package private global const?

Reply:
I don't particularly want to scope it wider than this function since it's not useful for anything else; what would be the benefit of this?

Reply:
The benefit would be not re-alloc'ing the same memory on every execution of the func.

if we're not gonna createt a global var, could we just resort to a bunch of if/else branches instead? or a series of switch statements? (that is trade the memory for logic)

Reply:
Turns out it doesn't make any appreciable difference in the benchmark to do one or the other (also you can't declare an array as `const`)
| method | time/op |
| -------- | --------- |
| declare in func 1 | 12549451 ns/op |
| declare in func 2 | 12351653 ns/op |
| declare in func 3 | 12411451 ns/op |
| declare in func avg | 12437518 ns/op |
| declare in global var 1 | 12406578 ns/op |
| declare in global var 2 | 12369299 ns/op |
| declare in global var 3 | 12417585 ns/op |
| declare in global var avg | 12397820 ns/op |

having a global var saves us 0.3%, which is pretty much in the noise with this low of a sample size :#